lib/Infer/AbstractInterpreter.cpp

```diff
@@ -390,35 +458,54 @@ namespace souper {
       // unit tests. Put minimum code outside it which you are sure of being
       // correct.
       if (isReservedConst(I->Ops[1]))
-	Result.Zero.setLowBits(1);
-      return getMostPreciseKnownBits(Result, BinaryTransferFunctionsKB::shl(KB0, KB1));
+        Result.Zero.setLowBits(1);
+      Result = getMostPreciseKnownBits(Result, BinaryTransferFunctionsKB::shl(KB0, KB1));
+      break;
     }
     case Inst::LShr : {
       if (isReservedConst(I->Ops[1]))
-	Result.Zero.setHighBits(1);
-      return getMostPreciseKnownBits(Result, BinaryTransferFunctionsKB::lshr(KB0, KB1));
+        Result.Zero.setHighBits(1);
+      Result = getMostPreciseKnownBits(Result, BinaryTransferFunctionsKB::lshr(KB0, KB1));
+      break;
     }
 //   case LShrExact:
 //     return "lshrexact";
     case Inst::AShr:
       if (isReservedConst(I->Ops[1])) {
-	if (KB0.Zero[KB0.getBitWidth() - 1])
-	  Result.Zero.setHighBits(2);
-	if (KB0.One[KB0.getBitWidth() - 1])
-	  Result.One.setHighBits(2);
+        if (KB0.Zero[KB0.getBitWidth() - 1])
+          Result.Zero.setHighBits(2);
```

Comment:
I don't understand this, is it saying that if we're going to synthesize a constant shift amount, we'll avoid synthesizing a zero there? and if so, is that true? please check with Zhengyang and make sure we never synthesize a zero

Reply:
yes, this is one of the major assumption used in a lot of places in dataflow pruning code. @manasij7479 

@zhengyangl we don't synthesize 0 for a reservedconst, right?

Reply:
Yes. The current constant synthesis can synthesis 0 if the enumeration reaches the guess (ashr x, c). If the NOP synthesis is turned off and the ashr is set to the only operator allowed, we get this synthesis result:

```
liuz@intel:~/work/build$ cat b.opt
%0:i8 = var
%1:i8 = add %0, 0:i8
%2:i8 = add %1, 0:i8
infer %2
liuz@intel:~/work/build$ ./souper-check -infer-rhs -souper-exhaustive-synthesis -z3-path=/usr/bin/z3 b.opt
; RHS inferred successfully
%3:i8 = ashr %0, 0:i8
result %3
```

For now we there is no code for setting constraints on the constants for synthesis. However, for this assumption, I think in the short term we can tell in our current cost model we never synthesize 0, based on the fact that bringing a new instruction will always increase the cost and the guesses without ashr NOP comes earlier when doing enumeration.

Reply:
We can synthesize 0.
It is rarely useful to do so.
If we can find a situation where it is useful to synthesize zero, we can fall back to consider that case specifically in the abstract interpreters.

Reply:
this code is ok *only* if you guys are certain that we never synthesize a zero. if we can synthesize a zero (I can't tell for sure from the discussion above) then it needs to be changed.

and if we keep this code, it needs a comment stating the assumption it is based on. even so, it's pretty dangerous and nasty.

Reply:
We can synthesize a zero, like in the example @zhengyangl has posted here.  
My argument is that there are no practical situations where synthesizing a zero produces an interesting RHS.

Reply:
if we can shift by zero then this is wrong, correct? ```Result.Zero.setHighBits(2);```

Reply:
If we prune using the assumption that the symbolic constant can not be zero, constant synthesis is avoided and we do not synthesize a zero.
The question here is: Is it a good idea to force synthesized constants to be non zero?
We discussed this a few months ago and came to the conclusion that it is a good idea.

Reply:
all I want is this static analysis to be (1) sound with respect to the code and (2) documented, since it relies on a pretty subtle assumption

Reply:
i have documented this assumption in the code now. marking it as resolved.